File: nodejs/src/access_token_common.js
import crypto from 'crypto'
import fs from 'fs';
import path from 'path'

import {ApiCommon} from './api_common.js'

export class AccessTokenCommon extends ApiCommon {

  constructor(api_config, {name = null}) {
    super();

    const auth = api_config.app_id + ':' + api_config.app_secret;
    const b64auth = Buffer.from(auth).toString('base64');
    this.api_config = api_config;
    this.api_uri = api_config.api_uri;
    this.auth_headers = {'Authorization': 'Basic ' + b64auth}
    if (name) {
      this.name = name;
    } else {
      this.name = 'access_token_' + api_config.version;
    }
    this.path = path.join(api_config.oauth_token_dir, this.name + '.json')
  }

  /**
   * This method tries to make it as easy as possible for a developer
   * to start using an OAuth access token. It fetches the access token
   * by trying all supported methods, in this order:
   *    1. Get from the process environment variable that is the UPPER CASE
   *       version of the this.name attribute. This method is intended as
   *       a quick hack for developers.
   *    2. Read the access_token and (if available) the refresh_token from
   *       the file at the path specified by joining the configured
   *       OAuth token directory, the this.name attribute, and the '.json'
   *       file extension.
   *    3. Execute the OAuth 2.0 request flow using the default browser
   *       and local redirect.
   */
  async fetch({scopes=null, refreshable=true}) {
    try {
      this.from_environment();
      return;
    } catch (err) {
      console.log(`reading ${this.name} from environment failed, trying read`);
      if (this.api_config.verbosity >= 3) {
        console.log('  because...', err);
      }
    }

    try {
      this.read();
      return;
    } catch (err) {
      console.log(`reading ${this.name} failed, trying oauth`);
      if (this.api_config.verbosity >= 3) {
        console.log('  because...', err);
      }
    }

    await this.oauth({scopes:scopes, refreshable:refreshable});
  }

  // constructor may not be async, so OAuth must be performed as a separate method.
  async oauth({scopes=null, refreshable=true}) {
    console.log('refresh() must be overriden.')
  }

  /**
   * Easiest path for using an access token: get it from the
   * process environment. Note that the environment variable name
   * is the UPPER CASE of the this.name instance attribute.
   */
  from_environment() {
    const env_access_token = process.env[this.name.toUpperCase()];
    if (env_access_token) {
      this.access_token = env_access_token;
      this.refresh_token = null;
    } else {
      throw 'No access token in the environment';
    }
  }

  /* Get the access token from the file at this.path. */
  read() {
    const data = JSON.parse(fs.readFileSync(this.path));
    this.name = data.name || 'access_token';
    const access_token = data.access_token;
    if (!access_token) {
      throw 'Access token not found in JSON file';
    }
    this.access_token = access_token;
    this.refresh_token = data.refresh_token;
    this.scopes = data.scopes;
    console.log(`read ${this.name} from ${this.path}`);
  }

  /* Store the access token in the file at this.path. */
  write() {
    const output = {'name': this.name,
                    'access_token': this.access_token,
                    'refresh_token': this.refresh_token,
                    'scopes': this.scopes
                   }
    const json = JSON.stringify(output, null, 2)
    /* Make credentials-bearing file as secure as possible with mode 0o600. */
    fs.open(this.path, 'w', 0o600, (err, fd) => {
      if (err) {
        throw 'Can not open file for write: ' + this.path;
      }
      fs.write(fd, json, (err, written, string) => {
        if (err) {
          throw 'Can not write file: ' + this.path;
        }
      })
    })
  }

  header(headers = {}) {
    headers['Authorization'] = 'Bearer ' + this.access_token;
    return headers;
  }

  hashed() {
    return crypto.createHash('sha256').update(this.access_token).digest('hex');
  }

  /**
   * Print the refresh token in a human-readable format that does not reveal
   * the actual access credential. The purpose of this method is for a developer
   * to verify when the refresh token changes.
   */
  hashed_refresh_token() {
    if (!this.refresh_token) {
      throw 'AccessToken does not have a refresh token';
    }
    return crypto.createHash('sha256').update(this.refresh_token).digest('hex');
  }

  async refresh() {
    console.log('refresh() must be overriden.')
  }
}

File: python/src/access_token_common.py
import base64
import hashlib
import json
import os
import pathlib

from api_common import ApiCommon

class AccessTokenCommon(ApiCommon):
    def __init__(self, api_config, name=None):

        if name:
            self.name = name
        else:
            self.name = 'access_token_' + api_config.version

        self.api_config = api_config
        self.path = pathlib.Path(api_config.oauth_token_dir) / (self.name + '.json')

        # use the recommended authorization approach
        auth = api_config.app_id + ':' + api_config.app_secret
        b64auth = base64.b64encode(auth.encode('ascii')).decode('ascii')
        self.auth_headers = {'Authorization': 'Basic ' + b64auth}

    def fetch(self, scopes=None, refreshable=True):
        """
        This method tries to make it as easy as possible for a developer
        to start using an OAuth access token. It fetches the access token
        by trying all supported methods, in this order:
           1. Get from the process environment variable that is the UPPER CASE
              version of the self.name attribute. This method is intended as
              a quick hack for developers.
           2. Read the access_token and (if available) the refresh_token from
              the file at the path specified by joining the configured
              OAuth token directory, the self.name attribute, and the '.json'
              file extension.
           3. Execute the OAuth 2.0 request flow using the default browser
              and local redirect.
        """
        try:
            self.from_environment()
            return
        except:
            print(f'reading {self.name} from environment failed, trying read')

        try:
            self.read()
            return
        except:
            print(f'reading {self.name} failed, trying oauth')

        self.oauth(scopes=scopes, refreshable=refreshable)

    def oauth(self, scopes=None, refreshable=True):
        print('oauth() must be overridden')

    def from_environment(self):
        """
        Easiest path for using an access token: get it from the
        process environment. Note that the environment variable name
        is the UPPER CASE of the self.name instance attribute.
        """
        self.access_token = os.environ[self.name.upper()]
        self.refresh_token = None

    def read(self):
        """
        Get the access token from the file at self.path.
        """
        with open(self.path, 'r') as jsonfile:
            data = json.load(jsonfile)
            self.name = data.get('name') or 'access_token'
            self.access_token = data['access_token']
            self.refresh_token = data.get('refresh_token')
            self.scopes = data.get('scopes')
        print(f'read {self.name} from {self.path}')

    def write(self):
        """
        Store the access token in the file at self.path.
        """
        with open(self.path, 'w') as jsonfile:
            # make credential-bearing file as secure as possible
            if 'chmod' in dir(os):
                os.chmod(jsonfile.fileno(), 0o600)
            # write the information to the file
            json.dump({'name': self.name,
                       'access_token': self.access_token,
                       'refresh_token': self.refresh_token,
                       'scopes': self.scopes},
                      jsonfile,
                      indent=2)

    def header(self, headers={}):
        headers['Authorization'] = 'Bearer ' + self.access_token
        return headers

    def hashed(self):
        """
        Print the access token in a human-readable format that does not reveal
        the actual access credential. The purpose of this method is for a developer
        to verify that the access token has changed after a refresh.
        """
        return hashlib.sha256(self.access_token.encode()).hexdigest()

    def hashed_refresh_token(self):
        """
        Print the refresh token in a human-readable format that does not reveal
        the actual access credential. The purpose of this method is for a developer
        to verify when the refresh token changes.
        """
        if not self.refresh_token:
            raise RuntimeError('AccessToken does not have a refresh token')
        return hashlib.sha256(self.refresh_token.encode()).hexdigest()

    def refresh(self):
        print('refresh() must be overridden.')
